src/main/java/org/cloudfoundry/promregator/cfaccessor/ReactiveCFPaginatedRequestFetcher.java

```diff
@@ -68,21 +115,59 @@ public ReactiveCFPaginatedRequestFetcher(InternalMetrics internalMetrics) {
 	 * The effort of implementing a key to lock-object mapping (as alternative) is expected to be rather high
 	 * and would imply the risk of memory leaks. 
 	 */
-	public <P, R> Mono<P> performGenericRetrieval(String retrievalTypeName, String logName, String key, R requestData,
+	public <P, R> Mono<P> performGenericRetrieval(RequestType requestType, String key, R requestData,
 			Function<R, Mono<P>> requestFunction, int timeoutInMS) {
+		final String retrievalTypeName = requestType.getMetricName();
+		final String logName = requestType.getLoggerSuffix();
+		
 		final String lock = (this.getClass().getCanonicalName()+"|"+retrievalTypeName+"|"+key).intern();
 		
 		synchronized (lock) {
 			Mono<P> result = null;
 
 			ReactiveTimer reactiveTimer = new ReactiveTimer(this.internalMetrics, retrievalTypeName);
+			
+			final Mono<P> enrichedRequestFunction = requestFunction.apply(requestData)
+				.timeout(Duration.ofMillis(timeoutInMS))
+				.retryBackoff(2, this.firstBackoffDelay, this.maxBackoffDelay);
+			/*
+			 * Note 1: Applying (i.e. calling) the function "requestFunction" here  
+			 * does not trigger the request to be sent to the CFCC. 
+			 * Instead, it just will create the corresponding Flux/Mono, which does
+			 * not have any subscriber yet. 
+			 * 
+			 * Note 2: There is a major difference between the coding modeled
+			 * 
+			 * requestFunction.apply(requestData).timeout(...)
+			 * 
+			 * and
+			 * 
+			 * someMono.flatMap(value -> requestFunction).timeout(...)
+			 * 
+			 * The major point here is that the first variant applies the timeout
+			 * only to the stream returned by requestFunction.apply(...), whilst
+			 * the second variant applies it to
+			 * 1. someMono,
+			 * 2. the flatMap function
+			 * 3. the return value of the requestFunction
+			 * 
+			 * The difference there is that in the second variant counting 
+			 * for the timeout starts already when someMono is subscribed to.
+			 * In the second variant, someMono is not considered.
+			 * 
+			 * In this case here, the difference may be huge: The first variant
+			 * puts a timeout on each request (which is what we want). The 
+			 * second variant means that timeout would be counting from the
+			 * first subscription happening - which is wrong especially in case 
+			 * of retry attempts.
+			 */
 
-			result = Mono.just(requestData)
+			result = this.rateLimitingMono(requestType).then(Mono.just(reactiveTimer))
 					// start the timer
-					.zipWith(Mono.just(reactiveTimer)).map(tuple -> {
-						tuple.getT2().start();
-						return tuple.getT1();
-					}).flatMap(requestFunction).timeout(Duration.ofMillis(timeoutInMS)).retry(2)
```

Comment:
With this change, moving the `retry` into timing changes this metric.  This metric indicated the 'speed' of the CFCC api. With this change, retries will be added to the duration, but not counted in the overall timer count.

I recommend keeping the previous timing approach.

Reply:
You are absolutely right. My bad.
Changed back with 5a6e806c11926253f445d5cad631e5cc7afc1d48